clients/src/main/java/org/apache/kafka/clients/consumer/internals/WakeupTrigger.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.clients.consumer.internals;
+
+import org.apache.kafka.common.KafkaException;
+import org.apache.kafka.common.errors.WakeupException;
+
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Ensures blocking APIs can be woken up by the consumer.wakeup().
+ */
+public class WakeupTrigger {
+    private AtomicReference<Wakeupable> pendingTask = new AtomicReference<>(null);
+
+    /*
+    Wakeup a pending task.  If there isn't any pending task, return a WakedupFuture, so that the subsequent call
+    would know wakeup was previously called.
+
+    If there are active taks, complete it with WakeupException, then unset pending task (return null here.
+    If the current task has already been wakedup, do nothing.
+     */
+    public void wakeup() {
+        pendingTask.getAndUpdate(task -> {
+            if (task == null) {
+                return new WakedupFuture();
```

Comment:
Do we ever clear a WakedupFuture from pendingTask if no active task is set afterward?

Reply:
The purpose setting a WakeupFuture when there isn't any active task is because we want the subsequent blocking call the return immediately.  Are you wondering if we should maintain this WakeupFuture forever? This is my intend at the moment.

Reply:
I was wondering when an application calls `Consumer.wakeup`. If it only calls when it intends to terminate, there is no need to clear WakedupFuture.

However, it seems that an application can call `Consumer.wakeup` to unblock a long `poll` call and then continue to make new `poll` calls. If we don't clear `WakedupFuture`, future blocking calls after the new `poll` call could fail unexpectedly?

Reply:
Hey Jun - For the blocking calls (syncCommit for example), the future should have been cleared in the finally block. I didn't implement wakeup for the poll method because it is not yet ready at the state of the current trunk.  (I'm not sure if i'm answering your question)

Basically - the wakeup future should be cleared in the finally blocks for the blocking APIs, see commitSync. Poll API is still incomplete so I didn't clear the future there.

Reply:
@philipnee : I was asking a slightly different question. Consider that an app calls `Consumer.wakeup` when there is no pending blocking operation. Then the app calls `Consumer.poll` before any blocking operation is called. At this point, a `WakedupFuture` is still left in `WakeupTrigger.pendingTask`. Some point later, a blocking operation is called. It will immediately throw an exception because of `WakedupFuture`, which seems unexpected.  So, should `Consumer.poll` clear `WakedupFuture` in `WakeupTrigger.pendingTask`?

Reply:
hey @junrao - sorry I misunderstood. In the example you provided, in fact, that is the current behavior.  It can be replicated in the integration test like this:

```
@Test
  def testWakeup(): Unit = {
    val consumer = createConsumer()
    val producer = createProducer()
    val numRecords = 10000
    val startingTimestamp = System.currentTimeMillis()
    sendRecords(producer, numRecords, tp, startingTimestamp = startingTimestamp)
    consumer.assign(List(tp).asJava)
    consumer.wakeup(); // wakeup w/o pending task
    assertThrows(classOf[WakeupException], () => consumer.poll(Duration.of(1000, java.time.temporal
      .ChronoUnit
        .MILLIS))) // throws immediately 
    println(consumer.poll(Duration.of(1000, java.time.temporal.ChronoUnit.MILLIS))) // prints some records
  }
```

In short - consumer.poll shouldn't clear WakeupFuture. In the Javadoc, it also didn't make an exception for the poll call see:

`...the next call to such a method will raise it instead.`

```
Wakeup the consumer. This method is thread-safe and is useful in particular to abort a long poll. The thread which is blocking in an operation will throw [WakeupException](https://kafka.apache.org/31/javadoc/org/apache/kafka/common/errors/WakeupException.html). If no thread is blocking in a method which can throw [WakeupException](https://kafka.apache.org/31/javadoc/org/apache/kafka/common/errors/WakeupException.html), the next call to such a method will raise it instead.
```

Reply:
Sorry @junrao - might have answered the wrong question (sorry, i was too eager to answer...).  But yes, poll should clear the pending task when the poll is completed, but currently, the poll isn't functioning at the moment, so I didn't clear the pending task there.  For the sake of clarity, do you think I should leave a comment/todo to clarify the intent.

Reply:
Thanks for the explanation, @philipnee. Yes, it would be useful to add a TODO comment to make this clear.